pkg/generator/asset_generator.go

```diff
@@ -0,0 +1,363 @@
+package generator
+
+import (
+	"fmt"
+	"path/filepath"
+	"strconv"
+
+	"github.com/openshift/csi-operator/pkg/generated-assets"
+)
+
+// AssetGenerator generates assets for CSI driver operators.
+type AssetGenerator struct {
+	flavour         ClusterFlavour
+	operatorConfig  *CSIDriverGeneratorConfig
+	replacements    []string
+	generatedAssets *generated_assets.CSIDriverAssets
+	reader          AssetReader
+}
+
+type AssetReader func(assetName string) ([]byte, error)
+
+// NewAssetGenerator creates a new AssetGenerator.
+func NewAssetGenerator(
+	flavour ClusterFlavour,
+	operatorConfig *CSIDriverGeneratorConfig,
+	reader AssetReader) *AssetGenerator {
+	return &AssetGenerator{
+		flavour:        flavour,
+		operatorConfig: operatorConfig,
+		replacements: []string{
+			"${ASSET_PREFIX}", operatorConfig.AssetPrefix,
+			"${ASSET_SHORT_PREFIX}", operatorConfig.AssetShortPrefix,
+			"${DRIVER_NAME}", operatorConfig.DriverName,
+		},
+		generatedAssets: &generated_assets.CSIDriverAssets{},
+		reader:          reader,
+	}
+}
+
+// GenerateAssets generates the assets for the CSI driver operator.
+// No assets are saved to the filesystem, they are returned as a CSIDriverAssets struct.
+func (gen *AssetGenerator) GenerateAssets() (*generated_assets.CSIDriverAssets, error) {
+	if err := gen.generateController(); err != nil {
+		return nil, err
+	}
+	if err := gen.generateGuest(); err != nil {
+		return nil, err
+	}
+	return gen.generatedAssets, nil
+}
+
+func (gen *AssetGenerator) generateController() error {
+	gen.generatedAssets = &generated_assets.CSIDriverAssets{
+		ControllerStaticResources: make(map[string][]byte),
+	}
+	if err := gen.generateDeployment(); err != nil {
+		return err
+	}
+
+	if err := gen.generateMonitoringService(); err != nil {
+		return err
+	}
+
+	if err := gen.collectControllerStaticAssets(); err != nil {
+		return err
+	}
+
+	if err := gen.patchController(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Apply all controller patches in the generator config (CSIDriverGeneratorConfig.ControllerConfig.AssetPatches)
+func (gen *AssetGenerator) patchController() error {
+	for _, patch := range gen.operatorConfig.ControllerConfig.AssetPatches {
+		if !patch.ClusterFlavours.Has(gen.flavour) {
+			continue
+		}
+		switch patch.GeneratedAssetName {
```

Comment:
The switch looks rather clumsy. I think it's worthy to have full mapping like `gen.generatedAssets.ControllerAssets[]` to have `generated_assets.ControllerDeploymentAssetName` included. Then `assetYAML := gen.generatedAssets.ControllerAssets[patch.GeneratedAssetName]` would work unconditionally.

And `gen.generatedAssets.ControllerStaticResources[]` could be a slice of that full mapping.

Reply:
+1 for clumsiness. I can add controller Deployment + node DaemonSet directly to Controller / NodeAssets easily, but it gets complicated with StorageClasses / VolumeSnapshotClasses below.

I added a separate patch moving everything just to two lists, `ControllerAssets` + `GuestAssets`, and detecting their `kind:` in runtime. This detection involves parsing all YAML files, possibly multiple times, but it should be used only during operator startup.